extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayCallbacksProcessor.java

```diff
@@ -0,0 +1,73 @@
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.flyway.runtime.FlywayBuildTimeConfig;
+
+class FlywayCallbacksProcessor {
+    private final Collection<String> dataSourceNames;
+    private final FlywayBuildTimeConfig flywayBuildConfig;
+    private final CombinedIndexBuildItem combinedIndexBuildItem;
+    private final BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer;
+
+    private FlywayCallbacksProcessor(Collection<String> dataSourceNames, FlywayBuildTimeConfig flywayBuildConfig,
+            CombinedIndexBuildItem combinedIndexBuildItem, BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer) {
+        this.dataSourceNames = dataSourceNames;
+        this.flywayBuildConfig = flywayBuildConfig;
+        this.combinedIndexBuildItem = combinedIndexBuildItem;
+        this.reflectiveClassProducer = reflectiveClassProducer;
+    }
+
+    public static FlywayCallbacksProcessor of(Collection<String> dataSourceNames, FlywayBuildTimeConfig flywayBuildConfig,
+            CombinedIndexBuildItem combinedIndexBuildItem, BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer) {
+        return new FlywayCallbacksProcessor(dataSourceNames, flywayBuildConfig, combinedIndexBuildItem,
+                reflectiveClassProducer);
+    }
+
+    public Map<String, Collection<Callback>> getCallbacks() throws Exception {
+        final Map<String, Collection<Callback>> callbacks = new HashMap<>();
+        for (String dataSourceName : dataSourceNames) {
+            final Collection<Callback> instances = callbacksForDataSource(dataSourceName);
+            callbacks.put(dataSourceName, instances);
+        }
+        return callbacks;
+    }
+
+    private Collection<Callback> callbacksForDataSource(String dataSourceName)
+            throws Exception {
+        final Optional<List<String>> callbackConfig = flywayBuildConfig.getConfigForDataSourceName(dataSourceName).callbacks;
+        if (!callbackConfig.isPresent()) {
+            return Collections.emptyList();
+        }
+        final Collection<String> callbacks = callbackConfig.get();
+        final Collection<Callback> instances = new ArrayList<>();
+        for (String callback : callbacks) {
+            final ClassInfo clazz = combinedIndexBuildItem.getIndex().getClassByName(DotName.createSimple(callback));
+            Objects.requireNonNull(clazz, "Flyway callback not found " + callback);
+            if (Modifier.isAbstract(clazz.flags()) || !clazz.hasNoArgsConstructor()) {
+                throw new IllegalArgumentException(
+                        "Invalid Flyway callback. It shouldn't be abstract and must have a default constructor");
+            }
+            final Class<?> clazzType = Class.forName(callback, false, Thread.currentThread().getContextClassLoader());
+            final Callback instance = (Callback) clazzType.getConstructors()[0].newInstance();
```

Comment:
I am pretty sure we'll need to register the constructors of these class for reflection if they are to work in native mode.

Furthermore, it probably makes sense to provide some kind of actionable error message (as it is too easy for a user to mistype the FQCN).

Reply:
I am not sure if I understood the refelction part, I thought that using the `BuildProducer<ReflectiveClassBuildItem>` was enough for reflection. Could you please give me some details to fully understand what should be done to fully support native compilation

Reply:
I guess I had missed the call to `reflectiveClassProducer` that I now see in the code.

That is enough, nothing more needs to be done to register for reflection